App/Services/RpcController.cs

```diff
@@ -0,0 +1,248 @@
+    /// <exception cref="InvalidOperationException">Another operation is in progress</exception>
+    /// <exception cref="RpcOperationException">If the sending of the start command fails</exception>
+    /// <exception cref="VpnLifecycleException">If the service reports that the VPN failed to start</exception>
+    public Task StartVpn(CancellationToken ct = default);
+
+    /// <summary>
+    ///     Stop the VPN. If the VPN is already not running, this may have no effect.
+    /// </summary>
+    /// <exception cref="InvalidOperationException">Another operation is in progress</exception>
+    /// <exception cref="RpcOperationException">If the sending of the stop command fails</exception>
+    /// <exception cref="VpnLifecycleException">If the service reports that the VPN failed to stop</exception>
+    public Task StopVpn(CancellationToken ct = default);
+}
+
+public class RpcController : IRpcController
+{
+    private readonly ICredentialManager _credentialManager;
+
+    private readonly RaiiSemaphoreSlim _operationLock = new(1, 1);
+    private Speaker<ClientMessage, ServiceMessage>? _speaker;
+
+    private readonly RaiiSemaphoreSlim _stateLock = new(1, 1);
+    private readonly RpcModel _state = new();
+
+    public RpcController(ICredentialManager credentialManager)
+    {
+        _credentialManager = credentialManager;
+    }
+
+    public event EventHandler<RpcModel>? StateChanged;
+
+    public RpcModel GetState()
+    {
+        using var _ = _stateLock.Lock();
+        return _state.Clone();
+    }
+
+    public async Task Reconnect(CancellationToken ct = default)
+    {
+        using var _ = await AcquireOperationLockNowAsync();
+        MutateState(state =>
+        {
+            state.RpcLifecycle = RpcLifecycle.Connecting;
+            state.VpnLifecycle = VpnLifecycle.Stopped;
+            state.VisibleAgents.Clear();
+        });
+
+        if (_speaker != null)
+            try
+            {
+                await _speaker.DisposeAsync();
+            }
+            catch (Exception e)
+            {
+                // TODO: log/notify?
+                Debug.WriteLine($"Error disposing existing Speaker: {e}");
+            }
+
+        try
+        {
+            var client =
+                new NamedPipeClientStream(".", "Coder.Desktop.Vpn", PipeDirection.InOut, PipeOptions.Asynchronous);
+            await client.ConnectAsync(ct);
+            _speaker = new Speaker<ClientMessage, ServiceMessage>(client);
+            _speaker.Receive += SpeakerOnReceive;
+            _speaker.Error += SpeakerOnError;
+            await _speaker.StartAsync(ct);
+        }
+        catch (Exception e)
+        {
+            MutateState(state =>
+            {
+                state.RpcLifecycle = RpcLifecycle.Disconnected;
+                state.VpnLifecycle = VpnLifecycle.Stopped;
+                state.VisibleAgents.Clear();
+            });
+            throw new RpcOperationException("Failed to reconnect to the RPC server", e);
```

Comment:
I think this exception is uncaught on the initialization of the tray window, which probably will crash the app.

Reply:
We run it in the background during TrayWindow.ctor, so any exceptions will get ignored and don't crash the app.